feat(bundler): codesign nested code on macos

pewsheen opened the pull request:
<!--
Update "[ ]" to "[x]" to check a box

Please make sure to read the Pull Request Guidelines: https://github.com/tauri-apps/tauri/blob/dev/.github/CONTRIBUTING.md#pull-request-guidelines
-->

### What kind of change does this PR introduce?
<!-- Check at least one. If you are introducing a new binding, you must reference an issue where this binding has been proposed, discussed and approved by the maintainers. -->

- [ ] Bugfix
- [x] Feature
- [ ] Docs
- [ ] New Binding issue #___
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

### Does this PR introduce a breaking change?
<!-- If yes, please describe the impact and migration path for existing applications in an attached issue. -->

- [ ] Yes, and the changes were approved in issue #___
- [x] No

### Checklist
- [ ] When resolving issues, they are referenced in the PR's title (e.g `fix: remove a typo, closes #___, #___`)
- [x] A change file is added if any packages will require a version bump due to this PR per [the instructions in the readme](https://github.com/tauri-apps/tauri/blob/dev/.changes/readme.md).
- [ ] I have added a convincing reason for adding this feature, if necessary

### Other information

Related: #8075 #8173 

References:
- [Code Signing Tasks](https://developer.apple.com/library/archive/documentation/Security/Conceptual/CodeSigningGuide/Procedures/Procedures.html#//apple_ref/doc/uid/TP40005929-CH4-SW2)

Notes:
- This patch does not cover all the cases but could be a start that we can sign nested code.
- Only signs codes inside standard location (`/MacOS`, `/Frameworks`, `/PlugIns`, `/XPCServices`, `/Helpers`)
- The risk of the patch is that if we are trying to sign something that is not signable, then the whole process will be stopped.
- Successfully codesign and notarized the reproduced repo from #8173 and the `test.framework` in tauri testing.

pewsheen committed:
feat(bundler): codesign nested code on macos

pewsheen committed:
Merge branch '1.x' into feat/codesign-r

pewsheen committed:
chore: update changelog tag

lucasfernog commented:
I think we should take this one step forward and apply this logic to the whole .app bundle instead, so it also signs user resources and [macOS custom files on v2](#7798).

pewsheen commented:
> I think we should take this one step forward and apply this logic to the whole .app bundle instead, so it also signs user resources and [macOS custom files on v2](#7798).

After doing some article search, I think I was just reimplementing the `codesign --deep`, which was [deprecated for reasons](https://developer.apple.com/forums/thread/129980).

Maybe we should take other approaches to sign those files, like providing a `sign_files` field in `tauri.config.json` and letting the user fill the path, entitlements, and is_executable while we can still auto codesign the files that generate from tauri itself.

hjmallon commented:
I'm really interested in solving this issue (probably in 1.x branch since it is for a live project). Would we rather have code signing on macOS:

1. Where the main app is automatically signed as currenlty, but all ofher files need a config line like @pewsheen says?
2. Something that does a `--deep` equivalent on everything, which is not very fine grained as mentioned?
3. Something that does a `--deep` equivalent on everuthing, but with a new config key for arguments for non standard parts?
4. Something else?

lucasfernog committed:
Merge remote-tracking branch 'origin/1.x' into feat/codesign-r

lucasfernog committed:
typo

lucasfernog committed:
also sign stuff in the Libraries folder

tested this for spacedrive, which has a bunch of dylib inside the libraries folder

lucasfernog commented on .changes/mac-bundler-nested-code-sign.md:
```diff
@@ -0,0 +1,6 @@
+"tauri-bundler": patch:feat
+---
+
+On macOS, support for signing nested code under the framework.
```
```suggestion
On macOS, support for signing nested .dylib, .app, .xpc and .framework under predefined directories inside the bundled frameworks ("MacOS", "Frameworks", "Plugins", "Helpers", "XPCServices" and "Libraries").
```

lucasfernog committed:
Update .changes/mac-bundler-nested-code-sign.md [skip ci]

raphaelmenges commented:
> Maybe we should take other approaches to sign those files, like providing a `sign_files` field in `tauri.config.json` and letting the user fill the path, entitlements, and is_executable while we can still auto codesign the files that generate from tauri itself.

Perhaps I overlook it in the documentation, but I am facing this exact issue and you could point me out to such `sign_files` field - if it exists. I am bundling a `.dylib` with my application that must be signed alongside the Tauri binaries to pass the notarization. Right now, notarization fails with `"The binary is not signed with a valid Developer ID certificate."` for that `.dylib`.

Worst case I can code sign the `.dylib` manually, yet, I of course prefer a Tauri-based solution.

therealpurplemana commented:
Did you find a solution to this? I'm running into this also.

raphaelmenges commented:
No, still a problem to my knowledge. I plan to use <https://github.com/lando/code-sign-action> to sign the `.dylib` manually. I would highly prefer to let the complete signing happen via Tauri.

therealpurplemana commented:
This isn't great.  It's causing issues with onnx and opencv libraries if you want to make release builds using them.  Isn't there a way to disable the unsafe library check and it still passes notarization?

raphaelmenges commented:
I found some workaround using the feature to add "frameworks" in the `tauri.conf.json`: https://github.com/tauri-apps/tauri/discussions/12001#discussioncomment-12417379

therealpurplemana commented:
> I found some workaround using the feature to add "frameworks" in the `tauri.conf.json`: [#12001 (comment)](https://github.com/tauri-apps/tauri/discussions/12001#discussioncomment-12417379)

This is exactly how I solved it also.  If you specify dylibs in Frameworks, Tauri will a) code sign it, b) properly @rpath it into the Frameworks folder.  I can confirm it passes Apple Notarization also and suspect even Gatekeeper.
